### PR TITLE
cpu/sam0_common: add hwrng driver

### PR DESCRIPTION
**Pull request by benpicco**
SAML1X, SAML2X and SAMD5X all share the same HWRNG peripheral.

This adds a driver for it.

### CODE REVIEW

**Comment by benpicco** [comment]
Hm, Murdock is confused

```
make: Entering directory '/tmp/dwq.0.8274662956598827/214d4a4189f5c8ea0f2c19669659fdab/tests/periph_hwrng'
There are unsatisfied feature requirements: periph_hwrng
\n\nEXPECT ERRORS!\n\n
Building application "tests_periph_hwrng" for "samr30-xpro" with MCU "saml21".
```

**Comment by dylad** [comment]
@benpicco I've open a ticket on Microchip support website to have an official answer from them. We should just wait a couple of days.

**Comment by benpicco** [comment]
So how should we continue? Just drop the RNG for the entire saml21 family?

**Comment by dylad** [comment]
@cladmi Does the Makefile changes look good to you ?

**Comment by cladmi** [comment]
Processing `Makefile.features` must be done before `Makefile.include`.
Changing this would have required a dedicated pull request and testing procedure.

Also murdock does not use `Makefile.include` for checking supported boards https://github.com/RIOT-OS/RIOT/blob/90e8da0a9ba23fcc43061693e7a05689f8f313c6/makefiles/info-global.inc.mk#L19
So when doing `make info-boards-supported` the `CPU_MODEL` is empty and does not go in this section.  Which I guess lead to the `There are unsatisfied feature requirements: periph_hwrng` and required `BLACKLISTING`.

The issue is that currently `CPU_MODEL` is defined in `Makefile.include` but is needed before. References:
* Makefile.features: prerequisites for moving CPU/CPU_MODEL to boards/Makefile.features #11478
* Tracking: move CPU/CPU_MODEL to Makefile.features #11477 
* Build dependencies - processing order issues #9913

Please give feedback over there if you are interested.



The current work around is to use `BOARD` until it is fixed with a comment. And I prefer whitelisting over blacklisting.

```
git grep '$(BOARD)' ':**/Makefile.features'
boards/common/arduino-atmega/Makefile.features:ifeq (,$(filter jiminy-mega256rfr2,$(BOARD)))
cpu/kinetis/Makefile.features:ifneq (,$(filter-out $(_KINETIS_BOARDS_WITHOUT_HWRNG),$(BOARD)))
cpu/stm32_common/Makefile.features:ifneq (,$(filter $(BOARDS_WITHOUT_HWRNG),$(BOARD)))
cpu/stm32f0/Makefile.features:ifeq (,$(filter nucleo-f031k6,$(BOARD)))
```

File: cpu/saml21/Makefile.features [hunk 1]
```diff
@@ -1 +1,6 @@
--include $(RIOTCPU)/sam0_common/Makefile.features
+# samr30 does not contain a TRNG
+ifeq (,$(findstring samr30,$(CPU_MODEL)))
```
Comment (cladmi) [inline cpu/saml21/Makefile.features]: Please replace it with whitelisting boards and  a comment on which CPU_MODEL have the feature instead of the opposite.

File: cpu/saml21/Makefile.features [hunk 2]
```diff
@@ -1 +1,6 @@
...
+  FEATURES_PROVIDED += periph_hwrng
+endif
+
+include $(RIOTCPU)/sam0_common/Makefile.features
```
Comment (cladmi) [inline cpu/saml21/Makefile.features]: This line should be a different commit/PR as it needs its specific review.

Nice find, I missed it when reviewing not included 'common' features.

**Comment by benpicco** [comment]
I've followed the precedent set by stm32 and kinetis and introduced a `BOARDS_WITHOUT_HWRNG` list that currently only contains samr30-xpro.

File: cpu/saml21/Makefile.features (hunk 2 above)
Comment (benpicco) [reply cpu/saml21/Makefile.features]: That line is not new, I just changed `-include` to `include` to be consistent with the rest of sam0.

File: cpu/saml21/Makefile.features (hunk 1 above)
Comment (benpicco) [reply cpu/saml21/Makefile.features]: stm32 and kinetis use a `BOARDS_WITHOUT_HWRNG` blacklist - I think this is easier to maintain.
I've added one to saml21 too.

File: cpu/saml21/Makefile.features (hunk 2 above)
Comment (cladmi) [reply cpu/saml21/Makefile.features]: Then please split it in another commit.

File: cpu/saml21/Makefile.features (hunk 1 above)
Comment (cladmi) [reply cpu/saml21/Makefile.features]: Defining the `BOARDS_WITHOUT_HWRNG` variable without clearing it in `makefiles/info-global.inc.mk` is somehow also wrong even if harmless.

But as it is currently done for the other family I agree.

**Comment by cladmi** [comment]
Please do not push force/rebase to master during a review. It somehow forces to re-review everything each time as it is a pain to track what changed.

File: cpu/saml21/Makefile.features (hunk 2 above)
Comment (benpicco) [reply cpu/saml21/Makefile.features]: Ok, I split it to another commit.

**Comment by benpicco** [comment]
Sorry, I thought it would be cleaner to drop the commit and force-push instead of pushing a revert commit.

**Comment by cladmi** [comment]
The current process is to not squash/rebase during a review as mentioned in https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#squash-commits-after-review

The makefiles changes are now good for me. @dylad I let you handle the rest.

**Comment by PeterKietzmann** [comment]
I've tested this PR on a saml11-xpro. First I ran *tests/periph_hwrng* which spits out "different" values. Then I ran *tests/rng* and its shell commands.

- Rate of 529 KiB/s seems ok (a remote-reva containing cc2538 has 507 KiB/s).
- FIPS 140-2 succeeds
- Bit distributions look equally distributed.

```
#  speed
# Running speed test, with seed 0 using HW RNG.
# 
# Running speed test for 10 seconds
# Collected 1355927 samples in 10.000037 seconds (529 KiB/s).
> fips
#  fips
# Running FIPS 140-2 test, with seed 0 using HW RNG.
# 
# - Monobit test: passed
# - Poker test: passed
# - Run test: passed
# - Longrun test: passed
> distributions
#  distributions
# Running distributions test, with seed 0 using HW RNG.
# 
# For 32-bit samples (min = 4907, max = 5064, avg = 4985):
# 00: ##########################################################################
# 01: ###########################################################################
# 02: ###########################################################################
# 03: ##########################################################################
# 04: ##########################################################################
# 05: ##########################################################################
# 06: ##########################################################################
# 07: ##########################################################################
# 08: ###########################################################################
# 09: #########################################################################
# 10: #########################################################################
# 11: #########################################################################
# 12: #########################################################################
# 13: ##########################################################################
# 14: ##########################################################################
# 15: ##########################################################################
# 16: ##########################################################################
# 17: #########################################################################
# 18: ###########################################################################
# 19: ##########################################################################
# 20: ###########################################################################
# 21: ##########################################################################
# 22: ###########################################################################
# 23: #########################################################################
# 24: ##########################################################################
# 25: #########################################################################
# 26: ##########################################################################
# 27: ##########################################################################
# 28: ##########################################################################
# 29: #########################################################################
# 30: ##########################################################################
# 31: ##########################################################################
```

**Comment by PeterKietzmann** [comment]
The code looks good to me too. I'll set all labels but "fundamentals" as I didn't participate in the initial discussion. @dylad please check and eventually remove your change request.

**Comment by dylad** [comment]
Same here, I'll take a look at it next week.

**Comment by dylad** [comment]
I just retest this PR on SAML10, SAML11, SAML21 & SAME54 with tests/periph_hwrng.

Output looks good to me !
